File: src/control/pure_pursuit/pure_pursuit/lifecycle_pure_pursuit.py
import rclpy
from rclpy.lifecycle import LifecycleNodeMixin, LifecyclePublisher, LifecycleState, TransitionCallbackReturn
from rclpy.node import Node
from rclpy.subscription import Subscription
from rclpy.timer import Timer

from ackermann_msgs.msg import AckermannDriveStamped
from driverless_msgs.msg import PathStamped
from sensor_msgs.msg import Image
from nav_msgs.msg import Path
from geometry_msgs.msg import PointStamped

from driverless_common.common import QOS_LATEST
from pure_pursuit.node_pure_pursuit import PurePursuit


class PurePursuitLifecycle(PurePursuit, LifecycleNodeMixin):
    """ """

    path_sub: Subscription
    control_pub: LifecyclePublisher
    debug_pub: LifecyclePublisher
    rvwp_publisher: LifecyclePublisher
    timer: Timer

    def __init__(self, node_name="pure_pursuit_lifecycle"):
        """
        Overrides __init__ in PurePursuit for lifecycle based pub/sub configuration
        """
        # initialise base node but not pure pursuit - so we dont get its params and pubs/subs
        Node.__init__(self, node_name)
        LifecycleNodeMixin.__init__(self, enable_communication_interface=True)

        self.initialise_params()

        if node_name == "pure_pursuit_lifecycle":
            self.get_logger().info("---Lifecycle Pure pursuit follower initialised---")

    def can_drive(self):
        if self.path is None:
            return False
        self.get_logger().info("Can drive", once=True)
        return True

    def on_configure(self, state: LifecycleState) -> TransitionCallbackReturn:
        self.get_logger().info("on_configure")
        self.control_pub = self.create_lifecycle_publisher(AckermannDriveStamped, "/control/driving_command", 10)
        self.rvwp_publisher = self.create_lifecycle_publisher(PointStamped, "/control/rvwp", 1)
        if self.DEBUG_IMG:
            self.debug_pub = self.create_lifecycle_publisher(Image, "/debug_imgs/pursuit_img", 1)
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state: LifecycleState) -> TransitionCallbackReturn:
        self.get_logger().info("on_activate")
        self.vel_max = self.get_parameter("vel_max").value
        self.vel_min = self.get_parameter("vel_min").value
        self.lookahead = self.get_parameter("lookahead").value

        # self.path_sub = self.create_subscription(PathStamped, "/planner/path", self.path_callback, QOS_LATEST)
        self.create_subscription(Path, "/planning/midline_path", self.path_callback, QOS_LATEST)
        self.timer = self.create_timer((1 / 50), self.timer_callback)
        return super().on_activate(state)

    def on_deactivate(self, state: LifecycleState) -> TransitionCallbackReturn:
        self.get_logger().info("on_deactivate")
        self.destroy_subscription(self.path_sub)
        self.destroy_timer(self.timer)
        return super().on_deactivate(state)

    def on_cleanup(self, state: LifecycleState) -> TransitionCallbackReturn:
        self.get_logger().info("on_cleanup")
        self.destroy_subscription(self.path_sub)
        self.destroy_timer(self.timer)
        self.destroy_lifecycle_publisher(self.control_pub)
        self.destroy_lifecycle_publisher(self.debug_pub)
        return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, state: LifecycleState) -> TransitionCallbackReturn:
        self.get_logger().info("on_shutdown")
        self.destroy_subscription(self.path_sub)
        self.destroy_timer(self.timer)
        self.destroy_lifecycle_publisher(self.control_pub)
        self.destroy_lifecycle_publisher(self.debug_pub)
        return TransitionCallbackReturn.SUCCESS

    def on_error(self, state: LifecycleState) -> TransitionCallbackReturn:
        self.get_logger().info("on_error")
        self.destroy_subscription(self.path_sub)
        self.destroy_timer(self.timer)
        self.destroy_lifecycle_publisher(self.control_pub)
        self.destroy_lifecycle_publisher(self.debug_pub)
        return TransitionCallbackReturn.SUCCESS


def main(args=None):  # begin ros node
    rclpy.init(args=args)
    node = PurePursuitLifecycle()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

File: src/control/pure_pursuit_cpp/src/lifecycle_pure_pursuit.cpp
#include <lifecycle_pure_pursuit.hpp>

PurePursuitLifecycle::PurePursuitLifecycle() :
        rclcpp_lifecycle::LifecycleNode("pure_pursuit_cpp_node", rclcpp::NodeOptions().use_intra_process_comms(false)) {
}

// Can't call shared_from_this() is a constructor :(
void PurePursuitLifecycle::initialise() {
    auto buffer = std::make_shared<tf2_ros::Buffer>(this->get_clock());
    pp = std::make_unique<PurePursuit>(shared_from_this(), buffer,
                                       std::make_shared<tf2_ros::TransformListener>(*buffer));
    RCLCPP_INFO(get_logger(), "---Pure Pursuit (C++) Node Initialised---");
}

PurePursuitLifecycle::CallbackReturn PurePursuitLifecycle::on_configure(rclcpp_lifecycle::State const &) {
    pp->driving_command_pub =
            this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("/control/driving_command", 10);

    RCLCPP_INFO(get_logger(), "On configure");
    return CallbackReturn::SUCCESS;
}

PurePursuitLifecycle::CallbackReturn PurePursuitLifecycle::on_activate(rclcpp_lifecycle::State const &) {
    pp->path_sub = this->create_subscription<nav_msgs::msg::Path>(
            "/planning/midline_path", QOS_LATEST, std::bind(&PurePursuit::path_callback, pp.get(), _1));
    pp->timer =
            this->create_wall_timer(std::chrono::milliseconds(20), std::bind(&PurePursuit::timer_callback, pp.get()));
    pp->driving_command_pub =
            this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>("/control/driving_command", 10);
    RCLCPP_INFO(get_logger(), "On activate");
    return CallbackReturn::SUCCESS;
}

PurePursuitLifecycle::CallbackReturn PurePursuitLifecycle::on_deactivate(rclcpp_lifecycle::State const &) {
    if (pp->path_sub) pp->path_sub.reset();
    if (pp->timer) pp->timer->reset();
    if (pp->driving_command_pub) pp->driving_command_pub.reset();
    RCLCPP_INFO(get_logger(), "On deactivate");
    return CallbackReturn::SUCCESS;
}

PurePursuitLifecycle::CallbackReturn PurePursuitLifecycle::on_cleanup(rclcpp_lifecycle::State const &) {
    if (pp->path_sub) pp->path_sub.reset();
    if (pp->timer) pp->timer->reset();
    if (pp->driving_command_pub) pp->driving_command_pub.reset();
    RCLCPP_INFO(get_logger(), "On cleanup");
    return CallbackReturn::SUCCESS;
}

PurePursuitLifecycle::CallbackReturn PurePursuitLifecycle::on_shutdown(rclcpp_lifecycle::State const &) {
    if (pp->path_sub) pp->path_sub.reset();
    if (pp->timer) pp->timer->reset();
    if (pp->driving_command_pub) pp->driving_command_pub.reset();
    RCLCPP_INFO(get_logger(), "On shutdown");
    return CallbackReturn::SUCCESS;
}

PurePursuitLifecycle::CallbackReturn PurePursuitLifecycle::on_error(rclcpp_lifecycle::State const &) {
    RCLCPP_INFO(get_logger(), "On error");
    return CallbackReturn::SUCCESS;
}

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PurePursuitLifecycle>();
    node->initialise();
    rclcpp::spin(node->get_node_base_interface());
    rclcpp::shutdown();
    return 0;
}
